src/clojure/nrepl/misc.clj

```diff
@@ -48,3 +53,23 @@
                                                (-> session meta :id)
                                                session)}))]
     (merge basis response)))
+
+(defn requiring-resolve
+  "Resolves namespace-qualified sym per 'resolve'. If initial resolve fails,
+  attempts to require sym's namespace and retries. Returns nil if sym could not
+  be resolved."
+  [sym]
+  (or (resolve sym)
+      (try
+        (require (symbol (namespace sym)))
+        (resolve sym)
+        (catch Exception _))))
+
+;; TODO: this should live in `nrepl.middleware.session`, but that would
+;; introduce a circular dependency with `nrepl.middleware.interruptible-eval`
+(defn interrupted?
```

Comment:
Probably we can just duplicate it in the two namespaces to avoid having this weird placement. But clearly there’s a bit of more work to be done to cleanly decouple the session and the eval middlewares. My main concern here is that we’re making a public API change here, that we’ll undo pretty quickly most likely. I guess no one would actually use this outside of nREPL, but we have to be careful with introducing/removing public API elements regardless.

Reply:
Yeah, duplicating it is better – updated to add the public version to `nrepl.middleware.session` and a private version to `nrepl.middleware.interruptible-eval`.